Use WP to enqueue assets, register translations [MAILPOET-5118] [MAILPOET-4930]

JanJakes opened the pull request:
## Description

To be able to use WP translations in all of our JS/TS codebase, we need to enqueue all scripts using the WP-native system.
For completeness, I added the same as well for CSS.

Further simplifications:
1. Removed `javascript` and `stylesheet` macros from Twig templates.
2. Removed `translations` blocks from Twig templates, `localize` now enqueues strings automatically.
3. Made `after_javascript` block work with WP-native queue.
4. Removed `after_css` block from Twig templates.

## Code review notes

_N/A_

## QA notes

Please, focus on the following areas:
1. General smoke check of all admin and public-facing pages to verify that JS and CSS loads.
2. Verify that translations work.
3. Do all the verification once without and once with the premium plugin.

## Linked PRs

https://github.com/mailpoet/mailpoet-premium/pull/793

## Linked tickets

[MAILPOET-5118]
[MAILPOET-4930]

## After-merge notes

Ping @JanJakes, please, to notify others about this change.

## Tasks

- [x] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [x] I added sufficient test coverage
- [x] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes


[MAILPOET-5118]: https://mailpoet.atlassian.net/browse/MAILPOET-5118?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[MAILPOET-4930]: https://mailpoet.atlassian.net/browse/MAILPOET-4930?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

JanJakes committed:
Extract admin pages assets setup out of form assets controller

[MAILPOET-5118]

JanJakes committed:
Enqueue all admin scripts using WP native functions

[MAILPOET-5118]

JanJakes committed:
Enqueue inline translations from Twig templates after mailpoet.js

[MAILPOET-5118]

JanJakes committed:
Remove translation blocks ("localize" now enqueues scrips directly)

[MAILPOET-5118]

JanJakes committed:
Render "after_javascript" block after other scripts

[MAILPOET-5118]

JanJakes committed:
Invoke "mailpoet_scripts_admin_before" from the asset controller

[MAILPOET-5118]

JanJakes committed:
Use WP logic to enqueue Parsley validation strings

[MAILPOET-5118]

JanJakes committed:
Inline analytics JS

[MAILPOET-5118]

JanJakes committed:
Remove Twig JS asset macro

[MAILPOET-5118]

JanJakes committed:
Enqueue all styles using WP native functions

[MAILPOET-5118]

JanJakes committed:
Remove "after_css" block

[MAILPOET-5118]

JanJakes committed:
Remove Twig CSS asset macro

[MAILPOET-5118]

JanJakes committed:
Fix issue with loadash overriding underscore on window._

Fixes: TypeError: this.activateMode is not a function
See also: https://github.com/mailpoet/mailpoet/pull/4977/files

[MAILPOET-5118]

JanJakes committed:
Call "mailpoet_styles_admin_after" from page renderer

[MAILPOET-5118]

johnolek commented on mailpoet/lib/AdminPages/AssetsController.php:
```diff
@@ -22,101 +23,99 @@ public function __construct(
-      Env::$assetsUrl . '/dist/js/' . $this->renderer->getJsAsset('commons.js'),
-      [],
+  private function enqueueJsEntrypoint(string $asset, array $dependencies = []): void {
+    $name = 'mailpoet_entrypoint';
```
I may not be understanding this correctly, but I'm a bit confused about why this is a variable instead of a constant.

JanJakes replied:
Fixed by https://github.com/mailpoet/mailpoet/commit/e0e06489231eed0e8b730de0b8ef21b9133940f4. I originally needed a constant name for all entrypoints due to different queuing logic I was playing with. In the current state it doesn't make much of a sense, so I reverted that.

johnolek commented on mailpoet/lib/WP/Functions.php:
```diff
@@ -33,6 +33,14 @@ public function doAction($tag, ...$args) {
   }
 
+  /**
+   * @param string $hook_name
```
PHPStan is mad about this:

```suggestion
   * @param string $hookName
```

johnolek replied:
Since this was a very simple change I just pushed a fix myself, 97e317d142ec6cab6ccea63cdc36ddf6b771da8d, to get the pipeline to run

johnolek reviewed (changes requested):
This looks great to me, thanks so much for your work on this!

Other than my one minor question, I noticed the readme still references the translations block, which no longer exists: https://github.com/mailpoet/mailpoet/blob/dceaafadc67b2b15fc1ee342bde5452f3bc40284/mailpoet/README.md#L187-L187

I think it would be good to update that line, and maybe take a look to ensure there isn't anything else that needs to be updated/removed (not that I saw anything else myself).

Edit: I hope I didn't jump the gun on reviewing this! I just noticed it wasn't actually assigned to me before.

JanJakes committed:
Fix timing of "mailpoet_scripts_admin_before" actioon trigger

[MAILPOET-5118]

JanJakes committed:
Update readme with current information about i18n usage

[MAILPOET-5118]

JanJakes committed:
Keep entrypoint asset name as script name, remove unnecessary check

[MAILPOET-5118]

JanJakes committed:
Enqueue Parsley translation strings in after "admin_vendor"

This is a better place and it corresponds to the current logic.

[MAILPOET-5118]

JanJakes commented:
@johnolek Thanks for the review!

> Edit: I hope I didn't jump the gun on reviewing this! I just noticed it wasn't actually assigned to me before.

Sorry, my bad. I already assigned it, but then noticed I have a wrong timing for the hook for premium JS.
I fixed it in https://github.com/mailpoet/mailpoet/pull/5190/commits/a233220ce3c8f0df98e9afd0add213f006fb506a, then also improved https://github.com/mailpoet/mailpoet/pull/5190/commits/49cf469942f7b28473d2d80eb890c3742245598a, and noticed the constant entrypoint asset name & check are unnecessary (https://github.com/mailpoet/mailpoet/pull/5190/commits/e0e06489231eed0e8b730de0b8ef21b9133940f4). I also updated readme (https://github.com/mailpoet/mailpoet/pull/5190/commits/bfa988cdef30bd200ae60300cf4af8c030fa6299).

johnolek commented:
@JanJakes Thanks for the updates, especially all the extra details you added to the readme!

Aschepikov reviewed (approved):
GJ!
